torch/csrc/jit/serialization/import.h

```diff
@@ -16,6 +17,35 @@ class ReadAdapterInterface;
 namespace torch {
 namespace jit {
 
+// used in torch.package deserialization
+class TORCH_API StorageContextTracker {
+ public:
+  explicit StorageContextTracker() {}
+
+  void addStorage(
+      const std::string& name,
+      c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl> storage) {
+    storage_map_.insert({name, storage});
+  }
+
+  bool hasStorage(const std::string& name) {
+    return storage_map_.find(name) != storage_map_.end();
+  }
+
+  c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl> getStorage(
+      const std::string& name) {
+    TORCH_INTERNAL_ASSERT(hasStorage(name));
+    return storage_map_.find(name)->second;
+  }
+  ~StorageContextTracker() = default;
+
+ private:
+  std::map<
+      std::string,
+      c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>>
+      storage_map_;
+};
```

Comment:
I chose to use a vector of `c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>>` instead of a vector of `c10::Storages` because the `c10::TensorImpl` holds the `c10::Storage` in a ref-counted way. I'm still trying to verify that this way of tracking the storages doesn't lead to memory leaks. The comment at the top of the `TensorImpl` class also made it seem like the correct choice: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L331

Reply:
I think c10::Storage is fine. It's a wrapper class that holds onto an underlying ref-counted StorageImpl. Let's use Storage here if we can to better capture intent.